server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -2731,7 +2734,11 @@ private boolean assertMaxSeqNoOfUpdatesIsAdvanced(Term id, long seqNo, boolean a
     private void restoreVersionMapAndCheckpointTracker(DirectoryReader directoryReader) throws IOException {
         final IndexSearcher searcher = new IndexSearcher(directoryReader);
         searcher.setQueryCache(null);
-        final Query query = LongPoint.newRangeQuery(SeqNoFieldMapper.NAME, getPersistedLocalCheckpoint() + 1, Long.MAX_VALUE);
+        final Query query = new BooleanQuery.Builder()
+            .add(LongPoint.newRangeQuery(
+                SeqNoFieldMapper.NAME, getPersistedLocalCheckpoint() + 1, Long.MAX_VALUE), BooleanClause.Occur.MUST)
+            .add(new DocValuesFieldExistsQuery(SeqNoFieldMapper.PRIMARY_TERM_NAME), BooleanClause.Occur.MUST)
```

Comment:
Can you add a comment saying that this skips nested docs?

Reply:
Can you use `Queries.newNonNestedFilter()` so that we can track the usage more easily ?

Reply:
Adjusted in https://github.com/elastic/elasticsearch/pull/51279/commits/bd0771ac782e6eca7953a9fc4b31a869a740755a.